# MSIB32500
## 1.	COURSE INFORMATION 

**Time:**	Saturday from 9:00AM – 12:00PM			
**Location:** Gleacher Center (room assignments are posted on the day; please check Graham screen in lobby or outside Graham office on lower level)		
**Units:** 	100 units 

* This course is an elective for all students in the Masters of Science in Biomedical Informatics. Open to other graduate students with permission from the MSc BMI program and course director.

* Important Note: Changes may occur to the syllabus at the instructor's discretion. When changes are made, students will be notified via Canvas and in-class announcement. More details for assignments will be provided at least one week prior to their due date.

## 2.	COURSE DESCRIPTION 
This course will follow on from the Introduction to Bioinformatics and include advanced bioinformatics analysis topics such as: Linux and high performance computing; R programming for bioinformatics; genomic data visualization; and RNA sequencing data analysis. The course will emphasize hands-on tutorials on how to perform analysis of gene expression microarrays, RNA-Seq data and integration of genomics and clinical data in lab sessions designed to reinforce lecture material.  

##3.	COURSE OBJECTIVES
By taking this course, students will learn how to perform advanced bioinformatics data analysis of high throughput genetics data using High Performance Computing Infrastructures.

## 4.	COURSE WORKLOAD

•	3 hours of lecture, class discussion, and in-class activities per week.
•	There is approximately 6 hours of outside class work assigned each week.

## 5.	GRADING POLICY
•	To allow students to practice and develop the skills learned in class, grading will be based on 5 homework assignments spread out thru the duration of the course. Homework assignments given one week are due by the beginning of next week session. When deemed necessary, the beginning of the next class or lab may include a review of common mistakes found on the homework from the previous week. Final grade will be an average of grades over the 5 homework assignments.   

## 6.	ATTENDANCE POLICY
This course involves regular hands-on use of bioinformatics tools in lab sessions in addition to more formal lectures.
•	At least 80% attendance is expected (not more than 2 absences for the quarter).
•	The Instructor would expect a notification via e-mail from students for the missed session.
•	Making up missed lectures could be done via recorded session.

## 7.	LATE WORK
•	Under special justified circumstances, students could be eligible for additional time to submit their homework up to two business days after the original deadline. Students should request this extension upon presenting evidence of especial/extraordinary circumstances (i.e. medical or family emergency).

## 8.	LAPTOP POLICY
MSc BMI is a program that focuses heavily on computer science and hands-on practice. Students are required to have a laptop available to bring to courses for in-class lab work. Students should not have laptops open during lecture time or in-class activities that are not related to computer usage. Please limit computer usage to note-taking; some instructors may request that laptops be closed for portions of class. This provides a more interactive and engaging experience for both students and the instructor.

## 9.	COURSE MATERIALS and SOFTWARE 

Reading assignments will be available on the Course Management System (Canvas).  

**Required:** 
- Developing Bioinformatics Computer Skills: An Introduction to Software Tools for Biological Applications by Cynthia Gibas and Per Jambeck. ISBN-13: 978-1565926646, ISBN-10: 1565926641.

- Statistics and Data Analysis for Microarrays Using R and Bioconductor, Second Edition by Sorin Drăghici. 
ISBN-13: 978-1439809754, ISBN-10: 1439809755.

**Recommended:** 
- Bioinformatics Data Skills: Reproducible and Robust Research with Open Source Tools 1st Edition, by Vince Buffalo, ISBN-10: 1449367372, ISBN-13: 978-1449367374
(Available at http://proquestcombo.safaribooksonline.com.proxy.uchicago.edu/book/bioinformatics/9781449367480)
- Introduction to Parallel Computing by Zbigniew J. Czech, ISBN-13: 978-1107174399, ISBN-10: 1107174392.

** 10.	COURSE SCHEDULE and WEEKLY REQUIRED READINGS

**Week 1: – Introduction to Linux for Bioinformatics** 

Session Leader or Guest Lecturer: Jorge Andrade, Ph.D.
Objectives:  Learn how the Linux file system is organized and how to navigate it; Learn to perform basic operations with directories and files in Linux; Learn how to transfer files between computers; Learn useful text searching, text extraction and text manipulation techniques, and its application in bioinformatics; Understand how to use Linux commands for data management and text manipulation in bioinformatics.

Required Reading: Developing Bioinformatics Computer Skills. Chapter 3, 4 and 5. 
Recommended: Bioinformatics Data Skills, Chapter 3, 4.

**Week 2 – Introduction to Parallel and distributed computing**
 
Session Leader or Guest Lecturer: Jorge Andrade, PhD
Objectives: Learn the basics of parallel and distributed computing environments, basic models for parallel computation, and design of parallel and distributed algorithms. 
Required Reading:  
Recommended: Introduction to Parallel Computing, chapters 1, 3 and 4; Applied Parallel Computing Springer series Lecture Notes in Computer Science. (LNCS). Volume 4699. pp 791-798 2007


**Week 3 - Using a High Performance Computing cluster for bioinformatics analysis**     
 
Session Leader or Guest Lecturer: Jorge Andrade, PhD 
Objectives: Learn how develop HPC batch scripts using the Linux shell to execute code and develop analysis pipelines to perform computationally expensive bioinformatics tasks. 
Required Reading: 
Recommended:

**Week 4 – Introduction to R programming**
 
Session Leader or Guest Lecturer: Jorge Andrade, PhD 
Objectives:  Develop R programming skills and learn how to use R for Bioinformatics analysis
Required Reading: Statistics and Data Analysis for Microarrays Using R and Bioconductor. Chapter 6 
Recommended:

**Week 5 – Elements of Statistics in R for gene expression analysis** 
 
Session Leader or Guest Lecturer: Jorge Andrade, PhD 
Objectives: Understand classical statistics approaches for the analysis of genomics data, including statistical hypothesis testing, analysis of variance and linear models.
Required Reading:  Statistics and Data Analysis for Microarrays Using R and Bioconductor. Chapter 12, 13 and 14.

**Week 6 - Analysis of Microarrays Using R and Bioconductor** 
 
Session Leader or Guest Lecturer: Jorge Andrade, PhD
Objectives:  Learn how to perform the analysis of DNA Microarray data using R and Bioconductor. 
Required Reading: Statistics and Data Analysis for Microarrays Using R and Bioconductor. Chapter 3, 14, 16, 19, 20 and 21.

**Week 7 – Genomics Data Visualization **
 
Session Leader or Guest Lecturer: Jorge Andrade, PhD
Objectives:  Learn how to visualize genomics data using a genome browser; Learn how to generate high-quality figures for publication (PCA, heatmap, sample/gene cluster, GO/pathways, etc.)
Required Reading: Statistics and Data Analysis for Microarrays Using R and Bioconductor. Chapter 17 and 18.

**Week 8 - Introduction to RNA-Seq**  
 
Session Leader or Guest Lecturer: Riyue Bao, PhD
Objectives: Understand how the RNA-Seq technology works; Learn analysis ‘best-practices’ and build an RNA-Seq analysis pipeline for raw data quality control and trimming, mapping reads to a reference genome, collect RNA-seq quality metrics and coverage statistics and quantification of transcript abundance.
Required Reading: TBD
Recommended: Review of “High-Throughput Sequencing Technologies “ from Mike Snyder’s group (http://www.sciencedirect.com.proxy.uchicago.edu/science/article/pii/S1097276515003408)


**Week 9 –  Differentially Gene Expression Identification using RNA-Seq**
 
Session Leader or Guest Lecturer: Riyue Bao, PhD
Objectives: Learn how to perform the downstream analysis of RNA-seq data; Detect genes differentially expressed between conditions; Identify pathways / network enriched in genes of interest.
Required Reading:  TBA
Recommended: 

**Week 10 –  Associating Gene Expression with Clinical Data

Session Leader or Guest Lecturer: Riyue Bao, PhD
Objectives: Learn the background and application of The Cancer Genome Atlas (TCGA) project; Learn the structure and access of Genomics Data Commons (GDC); Explore datasets hosted on GDC and Learn how to associate gene expression with clinical data.
Required Reading: TBA
Recommended:


## 11. Weekly Class Schedule

Week	Date	Session Leader/ Guest Lecturer	Topic	In-class activity	Assignment due
1 	4/1/17	Jorge Andrade			
2	4/8/17	Jorge Andrade			
3	4/15/17	Jorge Andrade			
4	4/22/17	Jorge Andrade			
5	4/29/17	Jorge Andrade			
6	5/6/17	Jorge Andrade			
7	5/13/17	Jorge Andrade			
8	5/20/17	Riyue Bao			
9	5/27/17	Riyue Bao			
10	6/3/17	Riyue Bao			

## 12.	COURSE EVALUATION

The Graham School administers web-based course evaluations to students for each course near the end of the quarter.  Your completion of both the Unit (course) and Faculty evaluations is required.  You will be sent the web-link and instructions via e-mail later in the quarter.  Your evaluation of the course and faculty is anonymous; your identity can not be linked with your responses.

## 13.	STUDENTS WITH DISABILITIES

In compliance with University of Chicago policy and equal access laws, we are available to discuss appropriate academic accommodations you may require as a student with a disability.  Request for academic accommodations need to be made during the first week of the quarter, except for unusual circumstances, so arrangements can be made.  Students are encouraged to register with Services for Students with Disabilities (SSD) for disability verification and for determination of reasonable academic accommodations.  

## 14.	ACADEMIC INTEGRITY AT UNIVERSITY OF CHICAGO

Students are expected to comply with University regulations regarding academic integrity.  If you are in doubt about what constitutes academic dishonesty, please speak to us before the assignment is due and/or examine the University’s Academic and Plagiarism Policy at http://studentmanual.uchicago.edu/policies

Academic dishonesty includes, but is not limited to cheating on an exam (e.g., copying others’ answers, providing information to others, using a crib sheet) or plagiarism of a paper (e.g., taking material from readings without citation, copying another student’s paper). Students are required to abide by University of Chicago’s academic integrity policy and failure to maintain academic integrity will likely result in a failing grade in the class and/or expulsion from the University.


## 15.	SEXUAL HARASSMENT POLICY

It is the policy of University of Chicago that no male or female member of the University of Chicago – students, faculty, administrators or staff – may sexually harass any other member of the community.  Sexual advances, requests for sexual favors, and other verbal or physical conduct of a sexual nature constitute harassment when: submission to such conduct is made or threatened to be made, either explicitly or implicitly, a term or condition of an individual’s employment or education; or submission to or rejection of such conduct is used or threatened to be used as the basis for academic or employment decisions affecting that individual; or such conduct has the purpose or effect of substantially interfering with an individual’s academic or professional performance or creating what a reasonable person would sense as an intimidating, hospital, or offensive  employment, educational, or living environment. 

For more information visit: [studentmanual.uchicago.edu](http://studentmanual.uchicago.edu/page/policy-harassment-discrimination-and-sexual-misconduct)
